AutoUpdate Framework

mbhavi opened the pull request:
This PR includes changes for Auto Update of Brackets.
The changes are with respect to:
1) Windows implementation
2) Mac implementation
3) A new UI

ping @swmitra @nethip for review

The Auto Update workflow along with the screenshots is described below : 

- **Notification of Update** : through current UpdateNotification dialog, which is triggered either from Help->CheckForUpdates or from the side bar button notification

![image](https://user-images.githubusercontent.com/34858846/38232411-80d35790-3734-11e8-9e63-c4c3314d7676.png)

- **Download of Update** : Once user clicks on GetItNow button, a download of the latest installer is triggered in the background, with a progress bar shown on the status bar.

![image](https://user-images.githubusercontent.com/34858846/38232577-2fea3122-3735-11e8-9dfa-cce96e148c25.png)

- **Download retrial in case of failure** : If the download fails, it is retried for a maximum of 5 attempts, and the same is reflected on the status bar UI.

![image](https://user-images.githubusercontent.com/34858846/38232867-41cc617a-3736-11e8-9bb6-c91318b25fbc.png)

- **Validation of the downloaded update** : Once download is successful, the SHA256 checksum is validated, and the info is displayed on status bar.

![image](https://user-images.githubusercontent.com/34858846/38232920-7f99bf70-3736-11e8-9467-9ad6efc8af9d.png)

- **Download Failure** : In case download fails, error is shown to the user towards bottom of the editor.

![image](https://user-images.githubusercontent.com/34858846/38232991-c00e02c8-3736-11e8-909d-f0199de65bf1.png)

- **Download and Validation success** : User is prompted with the update bar, with buttons for "Restart" and "Later" options.  Clicking Restart button would start the update procedure by quitting the app, and launching the installer. Clicking on Later button allows user to continue with the normal flow of Brackets.

![image](https://user-images.githubusercontent.com/34858846/38233213-63563cfc-3737-11e8-8bdb-da63be53000e.png)

- **Validation Failure** : If the validation of the downloaded update fails, a validation error is displayed to the user in update bar.

![image](https://user-images.githubusercontent.com/34858846/38233894-c4c849a6-3739-11e8-8966-889e4aeb306f.png)

- **Update in Progress** : Once the user clicks Restart button, the app is quit and installer gets launched.

![image](https://user-images.githubusercontent.com/34858846/38233299-b87a0916-3737-11e8-8c6c-f2cae01d195d.png)

- **Update Success** : If update is successful, new version of Brackets is launched automatically, with the success message on the update bar.

![image](https://user-images.githubusercontent.com/34858846/38233384-03ce875c-3738-11e8-9de7-c4996c39b9d2.png)

- **Update Failure** : If the update fails, the older version of Brackets is relaunched automatically, and the error message is displayed to the user in update bar, with a link to the brackets website for a manual update operation.

![image](https://user-images.githubusercontent.com/34858846/38233494-58934296-3738-11e8-8915-76a428e503c3.png)


Any **warnings** or errors encountered during the worflow are reflected in the update bar.
Screenshots for one of the warnings is attached here. It corresponds to the scenario where user clicked on Restart button, but had dirty files to which he/she chose Cancel in File Save dialog. In this case, the update will happen on the next quit of Brackets, and the same is displayed to the user as a warning on update bar.

![image](https://user-images.githubusercontent.com/34858846/38233687-041b60b2-3739-11e8-9530-a92d0954937b.png)

On MAC, instead of the installer, the update procedure is carried out by an update shell script. Rest everything is same for MAC, as above.

mbhavi committed:
AutoUpdate Framework : Initial Changes

abose commented:
Awesome! @mbhavi Can you add some screenshots of the new flow?

abose commented:
Is there a preference to disable auto updates?

abose commented:
Sometimes, users disable auto-updates to prevent breaking changes to their setup[extension compatibilities Et.Al.]. It would be good if it is configurable by a preference without having to remove the extension. 
Also is there an opt-in at first launch? or is the update process kicked off without confirmation from the user?
@petetnt @swmitra

mbhavi commented:
@abose, I have attached the screenshots and the step-by-step workflow description.
The update is kicked off only after the confirmation from the user as an input of GetItNow button click in UpdateNotification dialog.

mbhavi committed:
Addressing unit testing issues, error handling, code cleanup

mbhavi committed:
Merge branch 'master' into miglani/autoUpdate

mbhavi committed:
Merge branch 'master' of https://github.com/adobe/brackets into miglani/autoUpdate

mbhavi committed:
Clean up of code and spaces

mbhavi committed:
Error handling and code cleanup changes

nethip commented on src/document/DocumentCommandHandlers.js:
```diff
@@ -142,6 +142,17 @@ define(function (require, exports, module) {
+    var DIRTY_FILESAVE_CANCELLED = "dirtyFileSaveCancelled";
+
+    /**
+    Event triggered when error is returned when tried to check if Auto Update is progress
```
nit: `trying` instead of tried.

mbhavi replied:
Addressed this in 537430b

nethip commented on src/document/DocumentCommandHandlers.js:
```diff
@@ -848,6 +859,19 @@ define(function (require, exports, module) {
+     * Checks and handles if auto update is currently in progress
+     */
+    function checkIfAutoUpdateInProgress() {
+        brackets.app.isAutoUpdateInProgress(function (err, isAutoUpdateInProgress) {
```
Add `if (brackets.app.isAutoUpdateInProgress) {` to check if this function is registered or not.

mbhavi replied:
isAutoUpdateInProgress() function has been added to the core. It is not a registered function. Hence, we need not have this check.

mbhavi replied:
I strongly believe suppressing the error would turn out to be harmful in this case, not showing the error to the user would be misleading to him/her believing that the feature can work with just the brackets changes. 
Moreover, I don't see that with any of the other functions' invocations. I believe doing that will be unnecessary bloating of the code, as this is most likely to happen only in the dev machines, which I think should be okay.

nethip commented on src/document/DocumentCommandHandlers.js:
```diff
@@ -848,6 +859,19 @@ define(function (require, exports, module) {
+     */
+    function checkIfAutoUpdateInProgress() {
+        brackets.app.isAutoUpdateInProgress(function (err, isAutoUpdateInProgress) {
+            if (err === 0 && isAutoUpdateInProgress) {
```
No need for an explicit check, general rule we follow in `Brackets` is to check for error condition first. So your statements should be

```

if (err) {
 
} else (isAutoUpdateInProgress) {

}

```

mbhavi replied:
Addressed this in 537430b

nethip commented on src/document/DocumentCommandHandlers.js:
```diff
@@ -848,6 +859,19 @@ define(function (require, exports, module) {
+    function checkIfAutoUpdateInProgress() {
+        brackets.app.isAutoUpdateInProgress(function (err, isAutoUpdateInProgress) {
+            if (err === 0 && isAutoUpdateInProgress) {
+                CommandManager.trigger(exports.DIRTY_FILESAVE_CANCELLED);
```
Use `execute` over `trigger`.

nethip replied:
Help me understand. The name of the method is `checkIfAutoUpdateInProgress` but you are executing `exports.DIRTY_FILESAVE_CANCELLED` ? Shouldn't this be something like whoever is calling this method, should be returned a promise and the client can then act upon the promise state.

mbhavi replied:
I have now placed all the events related to Auto Update, needed in core, in UpdateNotification in order to ensure consistency.
Also, I now chose a better name for checkIfAutoUpdateProgress(). It is actually a handler function. It iinternally calls a function isAutoUpdateInProgress(), which returns a bool, and the wrapper function handles this bool. So, I have renamed the wrapper function to handleIfAutoUpdateInProgress(). 
Please refer to 537430b

nethip commented on src/document/DocumentCommandHandlers.js:
```diff
@@ -996,6 +1020,7 @@ define(function (require, exports, module) {
                     if (selectedPath) {
                         _doSaveAfterSaveDialog(selectedPath);
                     } else {
+                        checkIfAutoUpdateInProgress();
```
Could you explain (workflow) why this call is required here?

mbhavi replied:
This is needed for the workflow where user clicked Restart button for the update to happen, but had some dirty files, because of which the app could not be quit, and instead, he got the Unsaved Files dialog, This dialog has 3 buttons : Save/Don't Save/Cancel. If Don't Save is pressed, we quit the app and run the update. If Save is pressed and user saves the file successfully, then also we quit the app and update. But when the user chooses to either press Cancel button in the above dialog, or presses Save followed by Cancel in the Save As dialog, we show him a message saying that the update will now be reflected on relaunch, which inherently means that the update will now be carried out the next time the app quits.
So, for this, we need to know whether or not the save file was cancelled by either of the two scenarios mentioned here, so that we can prompt him with the required message.

nethip replied:
Thanks for explanation @mbhavi

nethip commented on src/extensions/default/AutoUpdate/StateHandler.js:
```diff
@@ -0,0 +1,205 @@
+            _file = FileSystem.getFileForPath(this.path);
+
+        _file.exists(function (err, exists) {
+            if (exists) {
```
Check for `err`.

mbhavi replied:
Addressed in 537430b

nethip commented on src/extensions/default/AutoUpdate/StateHandler.js:
```diff
@@ -0,0 +1,205 @@
+     */
+    function _write(filePath, json) {
+        var result = $.Deferred(),
+            _file = FileSystem.getFileForPath(filePath);
```
check for validity of `_file`

nethip replied:
I think there is not harm in checking for validity of `_file` given the fact that you are going to `stringify` the JSON irrespective of whether `_file` is valid or not.

mbhavi replied:
@nethip , the requirement here is to do a force write. If the file exists, the content is overwritten. If it doesn't exist, the file is created and written. I am passing the boolean to be "true" in writeText, which ensures this.
Hence, the validity is not needed here.

nethip replied:
@mbhavi Let us not have discussion of having a null check or not. It is a simple `null` check. If you still feel that is not required, post merging this PR, one of us will add it.

mbhavi replied:
@nethip , addressed in 4fb6582

nethip commented on src/extensions/default/AutoUpdate/UpdateInfoBar.js:
```diff
@@ -0,0 +1,167 @@
+            });
+
+            $restart.keyup(function (e) {
+                if (e.which === 32) { //32 is the keycode for space key
```
use a named variable for better readability.

mbhavi replied:
Addressed in  537430b

nethip commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,608 @@
+        updateDir = appSupportDirectory + '/updateTemp',
+        updateJsonPath = updateDir + '/' + 'updateHelper.json';
+
+    var updateJsonHandler = new StateHandler.StateHandler(updateJsonPath);
```
StateHandler.StateHandler?

mbhavi replied:
Addressed in  537430b

nethip commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,608 @@
+    functionMap["brackets.registerBracketsFunctions"] = registerBracketsFunctions;
+
+    UpdateNotification.on(UpdateNotification.GET_AUTOUPDATE_INSTALLER, handleGetNow);
+    CommandManager.on(DocumentCommandHandlers.DIRTY_FILESAVE_CANCELLED, dirtyFileSaveCancelled);
```
Refer to the comment that I put above in `DocumentCommandHandler` about generic command handlers.

mbhavi replied:
Please refer to my comments in DocumentCommandHandlers and UpdateNotification, where I have explained the use cases and requirements.

nethip commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,608 @@
+    });
+    updateDomain.on('data', receiveMessageFromNode);
+
+    AppInit.appReady(function () {
```
Check for platform and only then initialize AppInit.appReady as auto update is targeted for only Win and Mac.

mbhavi replied:
Addressed in  537430b

nethip commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,608 @@
+        UpdateInfoBar           = require("UpdateInfoBar");
+
+
+    var _modulePath = FileUtils.getNativeModuleDirectoryPath(module),
```
Execute the steps only for `Win` and `Mac`

mbhavi replied:
Addressed this in 537430b

nethip commented on src/extensions/default/AutoUpdate/node/AutoUpdateDomain.js:
```diff
@@ -0,0 +1,380 @@
+            []
+        );
+
+        domainManager.registerEvent(
```
Empty function?

mbhavi replied:
This is actually an event registration, for which no naming is required.

nethip replied:
Got it! Thanks

nethip commented on src/htmlContent/update-dialog.html:
```diff
@@ -1,17 +1,17 @@
 <div class="update-dialog modal">
     <div class="modal-header">
+        <img class="update-icon" src="styles/images/update_large_icon.svg">
```
Check the validity of the changed HTML on `linux`.

mbhavi replied:
@ayushrathi15 , please validate this in the linux builds.

nethip commented on src/utils/UpdateNotification.js:
```diff
@@ -31,6 +31,7 @@ define(function (require, exports, module) {
     var Dialogs              = require("widgets/Dialogs"),
```
Can all the changes made here be moved to the extension?

mbhavi replied:
The changes made here for AutoUpdate feature heavily use the update information like version info, URLs etc., which is maintained in this file. Hence, the changes should not be moved to the extension.

nethip replied:
@mbhavi One of the main reasons, we move most of the new code to an extension is to leverage pluggability model. So imagine, `AutoUpdate` extension is removed from the list of default extensions, the core should function fine and preferably with other functionality, which in this case is just open the `brackets.io` page. You need to follow something like what @swmitra has done with https://github.com/adobe/brackets/pull/14140. Have a default implementation, which an extension can override and provide implementation on how to get the latest update.

nethip commented:
I have reviewed this PR . Please look at the comments and incorporated the changes wherever required.

Also make sure that for `linux` no new code, added by this PR, is getting called. Make sure the new update workflow is for `win` and `mac` only and make changes appropriately, either in `HTML` or in the extension code.

Also it is a good idea to have a key to disable this feature. Please add a preference to disable auto update.

mbhavi committed:
Addressing Review comments

mbhavi commented on src/document/DocumentCommandHandlers.js:
```diff
@@ -1217,6 +1242,7 @@ define(function (require, exports, module) {
             )
                 .done(function (id) {
                     if (id === Dialogs.DIALOG_BTN_CANCEL) {
+                        checkIfAutoUpdateInProgress();
```
I am indeed sending the notification of a Cancel click on dirty file save operation. Please refer to my comments above.

mbhavi commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,608 @@
+
+    /**
+     * Checks if it is the first iteration of download
+     * @returns {boolean} - true if first iteration, false if is a retrial of download
```
Addressed in  537430b

mbhavi commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,608 @@
+     *                            to be called in case of
+     *                            successful setting of update state
+     */
+    function setUpdateStateInJSON(key, value, callback) {
```
This is indeed a promise based function. The same promise handling code snippet was getting used at multiple places, and so, in order to avoid the repetition of code, I have templatized the code snippet.

mbhavi commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,608 @@
+     */
+    function handleSafeToDownload() {
+
+        var downloadCB = function() {
```
Please refer to my comment above.

mbhavi committed:
Code cleanup

mbhavi committed:
Code cleanup, for event dispatching and handling

mbhavi commented:
@nethip , I have addressed and committed all the required changes. Also, I have replied to all the comments.
The PR is ready for review again.

nethip commented:
@mbhavi I think the preference to disable is still pending.

nethip commented:
@mbhavi I don't think the changes which I had proposed with `DocumentCommandHandler.js` are incorporated. 

There should be no reference of `AutoUpdate` anywhere inside `DocumentCommandHandler.js`. You could still borrow the mechanism but no reference to auto update. I still see methods like `handleIfAutoUpdateInProgress` or  in the PR. I am not sure if there is any any strong reason for hard wiring auto update into `DocumentCommandHandler.js`, as you could create new events and emit those events from here. The `AutoUpdate` extension can then register for those events and react accordingly..

nethip commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,630 @@
+    UpdateNotification.on(UpdateNotification.DIRTY_FILESAVE_CANCELLED, dirtyFileSaveCancelled);
+    UpdateNotification.on(UpdateNotification.AUTOUPDATE_ERROR, handleAutoUpdateError);
+
+    if(brackets.platform !== "linux") {
```
How about checking for the shell functions existence like `setUpdateParamsAndQuit`, and if not present, just bail out?

nethip commented on src/utils/UpdateNotification.js:
```diff
@@ -333,6 +447,12 @@ define(function (require, exports, module) {
         var versionInfoUrl;
 
+        //AUTOUPDATE_UNITTESTING
+        var updateJSONLocalhostURL = PreferencesManager.get("updateJSONLocalhostURL");
```
This needs to be removed.

nethip commented on src/utils/UpdateNotification.js:
```diff
@@ -389,6 +509,16 @@ define(function (require, exports, module) {
+
+                        //If no checksum field is present then we're setting it to 0, just as a safety check,
+                        // although ideally this situation should never occur in releases post its introduction.
+                        var platformInfo = getPlatformInfo(),
```
If you want to retain this , you might want to move all of this to `notifyGetUpdateInstaller`

mbhavi committed:
Merge branch 'master' of https://github.com/adobe/brackets into miglani/autoUpdate

mbhavi committed:
Merge branch 'master' of https://github.com/adobe/brackets into miglani/autoUpdate

mbhavi committed:
Addressing review comments, adding feature-flag/preference for AutoUpdate, making AutoUpdate no-op for Linux, Replacing icons

mbhavi commented:
@nethip , the review comments have been addressed now. I have added the preference to enable/disable auto update, made AutoUpdate no-op on Linux, replaced the icons, and moved most of the core code onto extension side. Please find the changes in b50f8de
The changes are ready to be reviewed again.

nethip commented on src/utils/UpdateNotification.js:
```diff
@@ -61,6 +71,10 @@ define(function (require, exports, module) {
 
+    //Preference for hosting latest installer on a local server
+    //AUTOUPDATE_UNITTESTING
+    PreferencesManager.stateManager.definePreference("autoUpdate.testUrl", "string", "");
```
Do you want to still retain this preference?

mbhavi replied:
@nethip , Yes we need to retain this preference, as this will be very much needed for unit testing purposes.

nethip replied:
Please have this preference in a separate branch. We can't be shipping with test URLs preference.

mbhavi replied:
@nethip, I have removed the preference from this branch. I will probably include the changes in a separate branch then, for unit testing purposes.

nethip commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,833 @@
+    AppInit.appReady(function () {
+        
+        // Auto Update is supported on Win and Mac, as of now
+        if (brackets.platform !== "linux") {
```
I think one of the review comments got missed here. You need to check if `brackets.app.setUpdateParams` is registered or not. Only then run through the flow.

nethip commented:
Thanks for addressing the review comments @mbhavi . I have added more comments. Please address those comments and I think we are good to merge from the dev perspective.

mbhavi committed:
Addressing review comments, minor code cleanup

mbhavi commented:
@nethip, I have addressed the new comments in 4fb6582. Please review.

nethip commented:
@mbhavi One last check. What if the node domain fails? Can we figure that out and bail out in your extension?

mbhavi committed:
Handling the node domain failure scenarios

mbhavi commented:
@nethip , I have addressed the failure scenario where the node domain fails in b272289. Please have a quick review for the same.

ayushrathi15 commented:
LGTM.

@nethip @mbhavi I am done with the private build testing. We are good to go ahead with the merge.

mbhavi committed:
Adding update_download_url to config.json and brackets.config.dist.json

mbhavi committed:
Fixing eslint and jslint errors, cleaning up package.json, adding console log message on node domain failure

mbhavi commented on src/extensions/default/AutoUpdate/main.js:
```diff
@@ -0,0 +1,840 @@
+                 }
+             })
+             .fail(function() {
+                 return;
```
@nethip, addressed this in 9b89697.

dmittleman commented:
Just downloaded 1.13 and am reading discussion from early April.  Here is a **use case** as to why **you need a preference to disable auto updates**.

I teach at DePaul using Brackets as the editor for my in class demonstrations.  Brackets is installed on all student machines in our teaching labs.  We are a large college (maybe 4,000 or so students).  We currently have about 30 computer labs with Brackets installed on about 800 machines.  Our IT staff annually creates 20 or so different "images" for these computers (based on lab purpose and computer OS/capabilities.   The IT staff applies specific images to specific computers, and re-images the computers every week or so via an automated process to ensure the machines are in a known and usable state.

So, if an individual student is at a machine when the update message comes up and opts to update, then that machine is in a different state than expected for the rest of the week.  Some students sitting in a classroom lab may be on a different version of Brackets than the instructor.

Our IT staff certainly wants control over changes of state overriding what an individual student user might opt to do.   I don't think we are unique in this situation.

Please tell me this use case has been considered.
